### PR TITLE
ref(hybrid-cloud): Denormalize user avatars 2/3 

### PR DESCRIPTION
**Pull request by mikejihbe**
Denormalizing user avatars for hybrid cloud. This allows us to only run the filestore service on region silos and avoids cross region queries.

Broken out from https://github.com/getsentry/sentry/pull/45115.

Process will be:

* Add migration and model changes. Dual write to new columns. 
* Backfill <-- We are here
* Change serializer to new model

### CODE REVIEW

File: src/sentry/migrations/0385_backfill_user_avatars.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    User = apps.get_model("sentry", "User")
+    # UserAvatar = apps.get_model("sentry", "UserAvagtar")
+
+    for user in RangeQuerySetWrapperWithProgressBar(User.objects.all()):
+        avatar = user.avatar.first()
+        if avatar is None:
+            continue
+        user.avatar_type = avatar.avatar_type
+        if avatar.avatar_type == 1:
+            user.avatar_url = f"https://sentry.io/avatar/{avatar.ident}/"
```
Comment (dashed) [inline src/sentry/migrations/0385_backfill_user_avatars.py:19]: This migration would run for self hosted users. This new value might not work.

Comment (markstory) [reply src/sentry/migrations/0385_backfill_user_avatars.py:19]: @dashed is right. We should use `system.region-api-url-template` or `system.url-prefix` from options here instead as the base URL. The `system.region-api-url-template` requires the current region value though.

Using `generate_region_url()` to get the base URL is likely the better option here.

File: src/sentry/migrations/0385_backfill_user_avatars.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    dependencies = [
+        ("sentry", "0384_backfill_installation_ids"),
+    ]
+
+    operations = [
+        migrations.RunPython(
+            backfill_user_avatar,
+            reverse_code=migrations.RunPython.noop,
+            hints={"tables": ["auth_user"]},
```
Comment (markstory) [inline src/sentry/migrations/0385_backfill_user_avatars.py:44]:
```suggestion
            hints={"tables": ["auth_user", "sentry_useravatar"]},
```

**Comment by corps** [comment]
@mikejihbe Something is wrong with this PR, it's showing unrelated snuba commits that may be code deleted from master?

File: tests/sentry/search/events/builder/test_query_builder.py [hunk 3]
```diff
@@ -0,0 +1,777 @@
...
+from sentry.exceptions import InvalidSearchQuery
+from sentry.search.events import constants
+from sentry.search.events.builder import QueryBuilder
+from sentry.testutils.cases import TestCase
+from sentry.utils.snuba import Dataset, QueryOutsideRetentionError
+
+pytestmark = pytest.mark.sentry_metrics
+
+
+class QueryBuilderTest(TestCase):
```
Comment (markstory) [inline tests/sentry/search/events/builder/test_query_builder.py:21]: Is this from a merge conflict/rebase?

File: src/sentry/migrations/0397_backfill_user_avatars.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+from sentry.api.utils import generate_region_url
+from sentry.new_migrations.migrations import CheckedMigration
+from sentry.utils.query import RangeQuerySetWrapperWithProgressBar
+
+
+def backfill_user_avatar(apps, schema_editor):
+    User = apps.get_model("sentry", "User")
+    # UserAvatar = apps.get_model("sentry", "UserAvagtar")
+
+    for user in RangeQuerySetWrapperWithProgressBar(User.objects.all()):
```
Comment (wedamija) [inline src/sentry/migrations/0397_backfill_user_avatars.py:14]: Could be worth using `prefetch_related` here to grab the avatar ahead of time

File: src/sentry/migrations/0397_backfill_user_avatars.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        avatar = user.avatar.first()
+        if avatar is None:
+            continue
+        user.avatar_type = avatar.avatar_type
+        if avatar.avatar_type == 1:
+            user.avatar_url = f"{generate_region_url()}/avatar/{avatar.ident}/"
+        user.save()
```
Comment (wedamija) [inline src/sentry/migrations/0397_backfill_user_avatars.py:21]: May as well just save in the `if` to avoid an unnecessary db call.

Comment (wedamija) [inline src/sentry/migrations/0397_backfill_user_avatars.py:21]: Also would be slightly better to use `update_fields` to only save the updated `avatar_url` field

File: tests/sentry/snuba/test_query_transform.py [hunk 6]
```diff
@@ -0,0 +1,2480 @@
+from datetime import datetime
+from unittest.mock import patch
```
Comment (markstory) [inline tests/sentry/snuba/test_query_transform.py:2]: Did this test come from a merge/rebase with master?

**Review by markstory — COMMENTED** [review]
I can get the extra tests cleaned up tomorrow morning.

**Review by markstory — APPROVED** [review]
Looks good to me. I ran the migration locally and checked that the generated URL can be fetched. In single-tenant environments the url-prefix will be used which is correct.

File: src/sentry/types/region.py [hunk 7]
```diff
@@ -147,12 +147,19 @@ def get_local_region() -> Region:
     """
     from django.conf import settings
 
+    from sentry.api.utils import generate_region_url
+
     if SiloMode.get_current_mode() == SiloMode.MONOLITH:
+        # In SAAS monolith mode (pre-region deployment), we use US region
+        address = generate_region_url()
+        if settings.SENTRY_SINGLE_ORGANIZATION:
+            address = "/"  # In ST or self-hosted, we rely on relative addresses
```
Comment (wedamija) [inline src/sentry/types/region.py:156]: Are these changes related to the migration?

Comment (markstory) [reply src/sentry/types/region.py:156]: Not directly, I can remove them from these changes.
